File: go/src/problem_010/problem_010.go
/*
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 * 
 * Find the sum of all the primes below two million.
 */

package main

import "fmt"

// const limit = 2000000
const limit = 100000

func sum(array []int) (s int) {
  for _, v := range array {
    s += v
  }
  return
}

func main() {
  sieve := new([limit - 2]int)
  for i := 2; i < limit; i++ {
    sieve[i - 2] = i
  }

  // Sieve of Eratosthenes
  for idx, val := range sieve {
    if val == 0 {
      continue
    }
    // fmt.Println("Sieving... ", val)

    for i := idx + 1; i < len(sieve); i++ {
      if sieve[i] % val == 0 {
        sieve[i] = 0
      }
    }
  }
  fmt.Println(sieve)
  fmt.Println(sum(sieve[0:]))
}

File: c/problem_010.c
/*
 * problem 10
 *
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 *
 * Find the sum of all the primes below two million.
 *
 * https://projecteuler.net/problem=10
 */

#include <euler.h>

#define MAX_PRIME 2000000

int main(void) {
  uint64_t sum = 0;
  mpz_t * prime_sieve = sieve_primes(MAX_PRIME);

  // Now that all the numbers have been sieved out, we can sum the entire sieve.
  for(int m = 2; m < MAX_PRIME; ++m) {
    if(mpz_tstbit(*prime_sieve, m)) sum += m;
  }
  printf("%lu\n", sum);
}

// Answer: 142913828922
